examples/vello_editor/src/text.rs

```diff
@@ -0,0 +1,249 @@
+        if self.selection.is_collapsed() {
+            let range = self.selection.focus().text_start..self.selection.focus().text_end;
+            ActiveText::FocusedCluster(&self.buffer[range])
+        } else {
+            ActiveText::Selection(&self.buffer[self.selection.text_range()])
+        }
+    }
+
+    pub fn handle_event(&mut self, event: &WindowEvent) {
+        match event {
+            WindowEvent::ModifiersChanged(modifiers) => {
+                self.modifiers = Some(*modifiers);
+            }
+            WindowEvent::KeyboardInput { event, .. } => {
+                if !event.state.is_pressed() {
+                    return;
+                }
+                let shift = self
+                    .modifiers
+                    .map(|mods| mods.state().shift_key())
+                    .unwrap_or_default();
+                match event.physical_key {
+                    PhysicalKey::Code(code) => match code {
+                        KeyCode::ArrowLeft => {
+                            self.selection = self.selection.prev_logical(&self.layout, shift);
+                        }
+                        KeyCode::ArrowRight => {
+                            self.selection = self.selection.next_logical(&self.layout, shift);
+                        }
+                        KeyCode::ArrowUp => {
+                            self.selection = self.selection.prev_line(&self.layout, shift);
+                        }
+                        KeyCode::ArrowDown => {
+                            self.selection = self.selection.next_line(&self.layout, shift);
+                        }
+                        KeyCode::Home => {
+                            self.selection = self.selection.line_start(&self.layout, shift);
+                        }
+                        KeyCode::End => {
+                            self.selection = self.selection.line_end(&self.layout, shift);
+                        }
+                        KeyCode::Delete => {
+                            let start = if self.selection.is_collapsed() {
+                                let range = self.selection.focus().text_start
+                                    ..self.selection.focus().text_end;
+                                let start = range.start;
+                                self.buffer.replace_range(range, "");
+                                start
+                            } else {
+                                self.delete_current_selection().unwrap()
+                            };
+                            self.update_layout(self.width, 1.0);
+                            self.selection = Selection::from_byte_index(&self.layout, start);
+                        }
+                        KeyCode::Backspace => {
+                            let start = if self.selection.is_collapsed() {
+                                let end = self.selection.focus().text_start;
+                                if let Some((start, _)) =
```

Comment:
```suggestion
                                // TODO: Handle Emoji, etc.
                                if let Some((start, _)) =
```

Reply:
Noted but I think this behavior should be an option. I believe some editors (mostly on desktop?) allow you to “decompose” emojis with backspace by removing modifiers. Mobile usually just deletes the whole thing for simplicity. 

Parley tracks emoji clusters so adding this here should be significantly simpler than the current code in masonry that tries to parse them manually.

Reply:
Hmm. Handy that we can use the existing data for this.

Decomposing also needs some care with the zero width joiners, of course.
I presumed that the decomposing behaviour I was seeing was a bug in vscode, but it does seem to be intentional.

Configuring would be good. Of course, it would also be possible to do even better here, but our job isn't to come up with new UX for emoji input/editing :)

Reply:
Decomposing emoji is highly dependent on specific editor features that I've never seen implemented in a useful way in a real application. On Android this could conceivably be handled by input methods (through corrections), but on desktop there is no standard UI for it.